src/rgw/driver/rados/topic_migration.cc

```diff
@@ -0,0 +1,288 @@
+                  + " with: " + cpp_strerror(r)
+                  + " expected format pubsub.{tenant}.bucket.{name}/{marker}!";
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  // migrate the notifications
+  rgw_pubsub_bucket_topics v1_bucket_topics;
+  rgw_bucket rgw_bucket_info(tenant, bucket_name);
+  rgw_bucket_info.marker = marker;
+  rgw::sal::RadosBucket rados_bucket(driver, rgw_bucket_info);
+  RGWObjVersionTracker bucket_topics_objv;
+  r = rados_bucket.read_topics(v1_bucket_topics, &bucket_topics_objv, y, dpp);
+  if (r == -ENOENT) {
+    return 0; // ok, someone else already migrated
+  }
+  if (r < 0) {
+    std::string s = "failed to read v1 bucket notifications from " + obj.to_str()
+                  + " with: " + cpp_strerror(r);
+    rgw_clog_warn(rados, s);
+    return r;
+  }
+
+  // in a for-loop that retries ECANCELED errors:
+  // {
+  // load the corresponding bucket by name
+  // break if marker doesn't match loaded bucket's
+  // merge with existing RGW_ATTR_BUCKET_NOTIFICATION topics (don't override existing v2)
+  // write RGW_ATTR_BUCKET_NOTIFICATION xattr
+  // }
+  std::unique_ptr<rgw::sal::Bucket> bucket;
+  r = -ECANCELED;
+  for (auto i = 0u; i < 15u && r == -ECANCELED; ++i) {
+    r = driver->load_bucket(dpp, rgw_bucket_info, &bucket, y);
+    if (r == -ENOENT) {
+      break; // bucket is deleted, we should delete the v1 notification
+    }
+    if (r < 0) {
+      std::string s = "failed to load the bucket from " + obj.to_str()
+                    + " with: " + cpp_strerror(r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+
+    if (bucket->get_marker() != marker) {
+      break;
+    }
+
+    rgw::sal::Attrs& attrs = bucket->get_attrs();
+
+    rgw_pubsub_bucket_topics v2_bucket_topics;
+    if (const auto iter = attrs.find(RGW_ATTR_BUCKET_NOTIFICATION); iter != attrs.end()) {
+      // bucket notification v2 already exists
+      try {
+        const auto& bl = iter->second;
+        auto biter = bl.cbegin();
+        v2_bucket_topics.decode(biter);
+      } catch (buffer::error& err) {
+        ldpp_dout(dpp, 1) << "ERROR: failed to decode v2 bucket topics for bucket: "
+                      << bucket->get_name() << dendl;
+        return -EIO;
+      }
+    }
+      
+    v2_bucket_topics.topics.merge(v1_bucket_topics.topics);
+    bufferlist bl;
+    v2_bucket_topics.encode(bl);
+    attrs[RGW_ATTR_BUCKET_NOTIFICATION] = std::move(bl);
+
+    r = bucket->merge_and_store_attrs(dpp, attrs, y);
+    if (r != -ECANCELED && r < 0) {
+      ldpp_dout(dpp, 1) << "ERROR: failed writing bucket instance info: " << cpp_strerror(-r) << dendl;
+      std::string s = "ERROR: failed writing bucket instance info: " + cpp_strerror(-r);
+      rgw_clog_warn(rados, s);
+      return r;
+    }
+  }
+
+  // delete v1 notification obj with Bucket::remove_topics()
+  r = rados_bucket.remove_topics(&bucket_topics_objv, y, dpp);
```

Comment:
i think there's an issue with the retry loop above. that loops up to 15 times as long as `r == -ECANCELED`

if we give up after 15 tries, this deletes the notification object as if its migration succeeded

i think we should check the value of `r` after this loop, and treat that as a fatal error instead